mlir/lib/Dialect/AIR/IR/AIRDialect.cpp

```diff
@@ -1108,6 +1109,390 @@ void WaitAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+    offsets.clear();
+    for (unsigned i = 0; i < input_ty.getRank(); i++)
+      offsets.push_back(constZero);
+  } else
+    return failure();
+
+  // Compose memref type as the memref propagates through the chain of memref
+  // ops.
+  for (auto memrefOp : memrefOpVec) {
+    if (auto transposeOp = dyn_cast<memref::TransposeOp>(memrefOp)) {
+      if (transposeOp.getPermutation().getNumInputs() != offsets.size())
+        continue;
+      offsets =
+          applyPermutationMap<Value>(transposeOp.getPermutation(), offsets);
+    } else if (auto expandShapeOp = dyn_cast<memref::ExpandShapeOp>(memrefOp)) {
+      // Init. offsets
+      for (int i = (int)expandShapeOp.getReassociationIndices().size() - 1;
+           i >= 0; i--) {
+        if (expandShapeOp.getReassociationIndices()[i].size() <= 1)
+          continue;
+        for (unsigned j = 1;
+             j < expandShapeOp.getReassociationIndices()[i].size(); j++)
+          offsets.insert(offsets.begin() + i,
+                         rewriter.create<arith::ConstantIndexOp>(loc, 0));
+      }
+    } else if (auto subviewOp = dyn_cast<memref::SubViewOp>(memrefOp)) {
+      if (subviewOp != memrefOpVec.front() && !subviewOp.hasZeroOffset())
+        subviewOp->emitOpError(
+            "is not the source op in a chain of memref layout transformation "
+            "ops, but applies a non-zero offset. This feature is NYI, and "
+            "leads to unexpected behaviour.");
+    }
+  }
+
+  // Memref type at sink memref op.
+  input_ty =
+      llvm::cast<MemRefType>(memrefOpVec.back()->getResultTypes().front());
+
+  // Compose sizes and strides from the output memref type's layout.
+  strides = extractStridesFromMemrefType(input_ty, rewriter);
+  sizes = extractSizesFromMemrefType(input_ty, rewriter);
+
+  return canonicalizeAIRDmaOperands(rewriter, offsets, sizes, strides,
+                                    input_ty);
+}
+
+//
+// Dma op
+//
+
+static LogicalResult
+ComposeMemrefOpOnDmaMemcpyNdSrc(DmaMemcpyNdOp op, PatternRewriter &rewriter) {
+
+  Value memref = op.getSrcMemref();
+  if (!memref)
+    return failure();
+  auto loc = op->getLoc();
+  Value input_memref;
+  SmallVector<Value> offsets, sizes, strides;
+  offsets = op.getSrcOffsets();
+  if (!offsets.empty())
+    return failure();
+  sizes = op.getSrcSizes();
+  if (!sizes.empty())
+    return failure();
+  strides = op.getSrcStrides();
+  if (!strides.empty())
+    return failure();
+
+  if (failed(ComposeMemrefOp(memref, rewriter, input_memref, offsets, sizes,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::DmaMemcpyNdOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), op.getDstMemref(),
+      op.getDstOffsets(), op.getDstSizes(), op.getDstStrides(), input_memref,
+      offsets, sizes, strides);
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
```

Comment:
Use the rewriter instead of direct IR modification:
```suggestion
  rewriter.replaceOp(op, newOp);
```

Reply:
Updated.